src/query/boolean_query/boolean_query.rs

```diff
@@ -19,6 +20,105 @@ use std::collections::BTreeSet;
+///        let mut index_writer = index.writer(3_000_000)?;
+///        index_writer.add_document(doc!(
+///            title => "The Name of the Wind",
+///        ));
+///        index_writer.add_document(doc!(
+///            title => "The Diary of Muadib",
+///        ));
+///        index_writer.add_document(doc!(
+///            title => "A Dairy Cow",
+///            body => "hidden",
+///        ));
+///        index_writer.add_document(doc!(
+///            title => "A Dairy Cow",
+///            body => "found",
+///        ));
+///        index_writer.add_document(doc!(
+///            title => "The Diary of a Young Girl",
+///        ));
+///        index_writer.commit().unwrap();
+///    }
+///
+///    let reader = index.reader()?;
+///    let searcher = reader.searcher();
+///
+///    // TermQuery "diary" must and "girl" must not be present
+///    let girl_term_query: Box<dyn Query> = Box::new(TermQuery::new(
+///        Term::from_field_text(title, "girl"),
+///        IndexRecordOption::Basic,
+///    ));
+///    let diary_term_query: Box<dyn Query> = Box::new(TermQuery::new(
+///        Term::from_field_text(title, "diary"),
+///        IndexRecordOption::Basic,
+///    ));
+///    let queries_with_occurs1 = vec![
+///        (Occur::Must, diary_term_query.box_clone()),
+///        (Occur::MustNot, girl_term_query),
+///    ];
+///    let query1 = BooleanQuery::from(queries_with_occurs1);
+///    let count1 = searcher.search(&query1, &Count)?;
+///    assert_eq!(count1, 1);
+///
+///    // BooleanQuery with 2 `TermQuery`s
+///    // "title:diary OR title:cow"
+///    let cow_term_query: Box<dyn Query> = Box::new(TermQuery::new(
+///        Term::from_field_text(title, "cow"),
+///        IndexRecordOption::Basic,
+///    ));
+///    let query2 = BooleanQuery::from(vec![
+///        (Occur::Should, diary_term_query.box_clone()),
+///        (Occur::Should, cow_term_query),
+///    ]);
+///    let count3 = searcher.search(&query2, &Count)?;
+///    assert_eq!(count3, 4);
+///
+///    // BooleanQuery comprising of subqueries of different types:
+///    // `TermQuery` and `PhraseQuery`
+///    // "title:diary OR \"dairy cow\""
+///    let phrase_query: Box<dyn Query> = Box::new(PhraseQuery::new(vec![
+///        Term::from_field_text(title, "dairy"),
+///        Term::from_field_text(title, "cow"),
+///    ]));
+///    let query4 = BooleanQuery::from(vec![
+///        (Occur::Should, diary_term_query.box_clone()),
+///        (Occur::Should, phrase_query),
+///    ]);
+///    let count4 = searcher.search(&query4, &Count)?;
+///    assert_eq!(count4, 4);
+///
+///    // ("title:diary OR \"dairy cow\"") AND body:found
+///    let body_query: Box<dyn Query> = Box::new(TermQuery::new(
+///        Term::from_field_text(body, "found"),
+///        IndexRecordOption::Basic,
+///    ));
+///    let query5 = BooleanQuery::from(vec![(Occur::Must, body_query),
+///                                         (Occur::Must, Box::new(query4))]);
```

Comment:
Here, I have had to Box query4 up to stop rustc from showing a type-error

```
error[E0308]: mismatched types
  --> src/query/boolean_query/boolean_query.rs:116:55
   |
92 |                                         (Occur::Must, query4)]);
   |                                                       ^^^^^^ expected struct `std::boxed::Box`, found struct `tantivy::query::BooleanQuery`
   |
   = note: expected type `std::boxed::Box<dyn tantivy::query::Query>`
              found type `tantivy::query::BooleanQuery`
```

I think I should expand the from method to deal with dynamic types, but I am not sure if it breaks any promises/interfaces and if it's a good API to give to everyone

Reply:
By dynamic type you mean dynamic trait?
The query needs ownership anyway, so we woudl have to box_clone. I'd rather be explicit.

Any solution that allows not to own the query will require introducing a lifetime or Generics or both.
I don't think we want that.